readonly_bs.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	// instantiate the block cache
+	blkLru, err := lru.New(maxBlocks)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for blocks: %w", err)
+	}
+
+	return &AllShardsReadBlockstore{
+		d:               d,
+		shardSelectF:    shardSelector,
+		blockstoreCache: bslru,
+		blockCache:      blkLru,
+	}, nil
+}
+
+func (ro *AllShardsReadBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, finalErr error) {
+	logbs.Debugw("bitswap Get called", "cid", c)
+	defer func() {
+		if finalErr != nil {
+			logbs.Debugw("bitswap Get: got error", "cid", c, "error", finalErr)
+		}
+	}()
+
+	mhash := c.Hash()
+	// do we have the block cached ?
+	if val, ok := ro.blockCache.Get(mhash.String()); ok {
+		logbs.Debugw("bitswap Get: returning from block cache", "cid", c)
+		return val.(blocks.Block), nil
+	}
+
+	// fetch all the shards containing the multihash
+	shards, err := ro.d.ShardsContainingMultihash(ctx, mhash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch shards containing the block: %w", err)
+	}
+	if len(shards) == 0 {
+		return nil, errors.New("no shards contain the requested block")
+	}
+
+	// do we have a cached blockstore for a shard containing the required block ? If yes, serve the block from that blockstore
+	for _, sk := range shards {
+		// a valid cache hit here updates the priority of the shard's blockstore in the LRU cache.
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
+
+		// add the block to the block cache
+		logbs.Debugw("bitswap Get: returning from block store cache", "cid", c)
+		ro.blockCache.Add(mhash.String(), blk)
+		return blk, nil
+	}
+
+	// ---- we don't have a cached blockstore for a shard that can serve the block -> let's build one.
+
+	// select a valid shard that can serve the retrieval
+	sk, err := ro.shardSelectF(c, shards)
+	if err != nil {
+		return nil, fmt.Errorf("failed to select a shard: %w", err)
+	}
+
+	// load blockstore for the selected shard and tru to serve the cid from that blockstore.
+	resch := make(chan ShardResult, 1)
+	if err := ro.d.AcquireShard(ctx, sk, resch, AcquireOpts{}); err != nil {
+		return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
+	}
+	var res ShardResult
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res = <-resch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, res.Error)
+		}
+	}
+
+	sa := res.Accessor
+	bs, err := sa.Blockstore()
+	if err != nil {
+		return nil, fmt.Errorf("failed top load read only blockstore for shard %s: %w", sk, err)
+	}
+
+	blk, err := bs.Get(ctx, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	// update the block cache and the blockstore cache
+	ro.blockstoreCache.Add(sk, &accessorWithBlockstore{sa, bs})
```

Comment:
I'm wondering if we need the accessorWithBlockstore struct. Can we just store the accessor in the blockstore cache and call Blockstore() on the accessor each time we need the blockstore? Or does accessor.Blockstore() actually create a new blockstore?

Reply:
Yeah, there's a bunch of work that needs to be done to get a blockstore from an accessor.

Open/mmap the CARv1 payload and combine it with the CARv2 Index to get the read only blockstore. We should ideally do this only once.